src/Features/CSharp/Portable/MoveToResx/CSharpMoveToResxCodeFixProvider.cs

```diff
@@ -0,0 +1,235 @@
+                    return match;
+            }
+            // 2. Look for .resx in the same folder with any name
+            if (inSameDir.Count > 0)
+                return inSameDir.First();
+
+            // 3. Look for .resx in parent folders (upward scan), prefer preferred names
+            var currentDir = docDir;
+            while (!string.IsNullOrEmpty(currentDir))
+            {
+                foreach (var preferred in preferredNames)
+                {
+                    var match = allResx.FirstOrDefault(f =>
+                        Path.GetDirectoryName(f.FilePath).Equals(currentDir, StringComparison.OrdinalIgnoreCase) &&
+                        Path.GetFileNameWithoutExtension(f.FilePath).Equals(preferred, StringComparison.OrdinalIgnoreCase));
+                    if (match != null)
+                        return match;
+                }
+                currentDir = Path.GetDirectoryName(currentDir);
+            }
+
+            // 4. Fallback: any .resx in the project
+            return allResx.FirstOrDefault();
+        }
+
+        //private static IEnumerable<TextDocument> GetResxAdditionalFiles(Project project)
+        //{
+        //    foreach (var file in project.AdditionalDocuments)
+        //    {
+        //        if (file.FilePath != null &&
+        //            string.Equals(".resx", Path.GetExtension(file.FilePath), StringComparison.OrdinalIgnoreCase))
+        //        {
+        //            yield return file;
+        //        }
+        //    }
+        //}
+
+        // This method parses, updates, and saves the resx file in memory, and replaces the string literal with a resource reference
+        private static async Task<Solution> MoveStringToResxAndReplaceLiteralAsync(
+            Document document,
+            LiteralExpressionSyntax stringLiteral,
+            TextDocument resxFile,
+            CancellationToken cancellationToken)
+        {
+            // 1. Get the SourceText from the resx AdditionalDocument
+            var resxSourceText = await resxFile.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            // 2. Parse the SourceText with XDocument
+            var xdoc = XDocument.Parse(resxSourceText.ToString());
+
+            // 3. Add or update the resource entry
+            var value = stringLiteral.Token.ValueText;
+            var name = ToDeterministicResourceKey(value);
+
+            var dataElement = xdoc.Root.Elements("data")
+                .FirstOrDefault(e => (string)e.Attribute("name") == name);
+
+            if (dataElement == null)
+            {
+                dataElement = new XElement("data",
+                    new XAttribute("name", name),
+                    new XAttribute(XNamespace.Xml + "space", "preserve"),
+                    new XElement("value", value));
+                xdoc.Root.Add(dataElement);
+            }
+            else
+            {
+                dataElement.Element("value")!.Value = value;
```

Comment:
there's no guarantee that the resx file will be formed in the way you want.  it's important to check for that and act gracefully there.  these are also good test cases.